wgengine/magicsock/magicsock.go

```diff
@@ -1639,38 +1560,73 @@ Top:
 		// Clos() on magicsock, and expects ReceiveIPv4 to unblock
 		// with an error so it can clean up.
 		return 0, nil, errors.New("socket closed")
+	case dm = <-c.derpRecvCh:
+		// Below.
 	}
-}
 
-func (c *Conn) ReceiveIPv6(b []byte) (int, conn.Endpoint, error) {
-	if c.pconn6 == nil {
-		return 0, nil, syscall.EAFNOSUPPORT
+	var regionID int
+	n, regionID = dm.n, dm.regionID
+	ncopy := dm.copyBuf(b)
+	if ncopy != n {
+		err = fmt.Errorf("received DERP packet of length %d that's too big for WireGuard ReceiveIPv4 buf size %d", n, ncopy)
+		c.logf("magicsock: %v", err)
+		return 0, nil, err
 	}
-	for {
-		n, pAddr, err := c.pconn6.ReadFrom(b)
-		if err != nil {
-			return 0, nil, err
-		}
-		addr := pAddr.(*net.UDPAddr)
-		ipp, ok := netaddr.FromStdAddr(addr.IP, addr.Port, addr.Zone)
-		if !ok {
-			continue
-		}
-		if stun.Is(b[:n]) {
-			c.stunReceiveFunc.Load().(func([]byte, netaddr.IPPort))(b[:n], ipp)
-			continue
-		}
-		if c.handleDiscoMessage(b[:n], ipp) {
-			continue
+
+	ipp := netaddr.IPPort{IP: derpMagicIPAddr, Port: uint16(regionID)}
+	if c.handleDiscoMessage(b[:n], ipp) {
+		return 0, nil, errLoopAgain
+	}
+
+	var (
+		didNoteRecvActivity bool
+		discoEp             *discoEndpoint
+		asEp                *addrSet
+	)
+	c.mu.Lock()
+	if dk, ok := c.discoOfNode[tailcfg.NodeKey(dm.src)]; ok {
+		discoEp = c.endpointOfDisco[dk]
+		// If we know about the node (it's in discoOfNode) but don't know about the
+		// endpoint, that's because it's an idle peer that doesn't yet exist in the
+		// wireguard config. So run the receive hook, if defined, which should
+		// create the wireguard peer.
+		if discoEp == nil && c.noteRecvActivity != nil {
+			didNoteRecvActivity = true
+			c.mu.Unlock()          // release lock before calling noteRecvActivity
+			c.noteRecvActivity(dk) // (calls back into CreateEndpoint)
+			// Now require the lock. No invariants need to be rechecked; just
+			// 1-2 map lookups follow that are harmless if, say, the peer has
+			// been deleted during this time.  In that case we'll treate it as a
+			// legacy pre-disco UDP receive and hand it to wireguard which'll
+			// likely just drop it.
```

Comment:
This comment is no longer correct. We're near certain to drop a packet if it falls into the legacy codepath.